### PR TITLE
Initialize type arguments for anonymous inner class that use diamonds.

### LINKED ISSUE #2779: Using diamond operator with anonymous classes leads to errors
Using the diamond operator with anonymous classes as in: 
```java
import org.checkerframework.checker.nullness.qual.Nullable;

public interface Handle<T> {
	
    T get();

    void set(T val);

    static <@Nullable T> Handle<T> immutable(T val) {
        return new Handle<>() {
            @Override
            public T get() {
                return val;
            }

            @Override
            public void set(T val) {
                throw new UnsupportedOperationException();
            }
        };
    }
}
```
causes crashes: 
```
error: AnnotatedDeclaredType's element has a different number of type parameters than type.
  type=Handle
  element=Handle
  Compilation unit: Handle.java
  Last visited tree at line 3 column 1:
  public interface Handle<T> {
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.javacutil.BugInCF.<init>(BugInCF.java:14)
  org.checkerframework.framework.type.SupertypeFinder$SupertypeFindingVisitor.visitDeclared(SupertypeFinder.java:154)
  org.checkerframework.framework.type.SupertypeFinder.directSuperTypes(SupertypeFinder.java:46)
  org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType.directSuperTypes(AnnotatedTypeMirror.java:929)
  org.checkerframework.framework.util.AnnotatedTypes.getSuperTypes(AnnotatedTypes.java:591)
  org.checkerframework.framework.util.AnnotatedTypes.overriddenMethods(AnnotatedTypes.java:614)
  org.checkerframework.framework.type.AnnotatedTypeFactory.inheritOverriddenDeclAnnos(AnnotatedTypeFactory.java:3319)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getDeclAnnotations(AnnotatedTypeFactory.java:3299)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getDeclAnnotation(AnnotatedTypeFactory.java:3229)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getDeclAnnotation(AnnotatedTypeFactory.java:3164)
  org.checkerframework.framework.util.ContractsUtils.getPreconditions(ContractsUtils.java:245)
  org.checkerframework.framework.flow.CFAbstractTransfer.addInformationFromPreconditions(CFAbstractTransfer.java:518)
  org.checkerframework.framework.flow.CFAbstractTransfer.initialStore(CFAbstractTransfer.java:258)
  org.checkerframework.framework.flow.CFAbstractTransfer.initialStore(CFAbstractTransfer.java:93)
  org.checkerframework.dataflow.analysis.Analysis.init(Analysis.java:470)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:183)
  org.checkerframework.framework.flow.CFAbstractAnalysis.performAnalysis(CFAbstractAnalysis.java:91)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1253)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:1171)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1520)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:258)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:320)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:167)
  jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
  jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:82)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:1012)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:522)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:515)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:182)
  jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:828)
  jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
  jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1418)
  jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1375)
  jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:973)
  jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:311)
  jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:170)
  jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:57)
  jdk.compiler/com.sun.tools.javac.Main.main(Main.java:43)
error: AnnotatedDeclaredType's element has a different number of type parameters than type.
  type=@Initialized @NonNull Handle
  element=Handle
  Compilation unit: Handle.java
  Last visited tree at line 3 column 1:
  public interface Handle<T> {
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.javacutil.BugInCF.<init>(BugInCF.java:14)
  org.checkerframework.framework.type.SupertypeFinder$SupertypeFindingVisitor.visitDeclared(SupertypeFinder.java:154)
  org.checkerframework.framework.type.SupertypeFinder.directSuperTypes(SupertypeFinder.java:46)
  org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType.directSuperTypes(AnnotatedTypeMirror.java:929)
  org.checkerframework.framework.util.AnnotatedTypes.getSuperTypes(AnnotatedTypes.java:591)
  org.checkerframework.framework.util.AnnotatedTypes.overriddenMethods(AnnotatedTypes.java:614)
  org.checkerframework.framework.type.AnnotatedTypeFactory.inheritOverriddenDeclAnnos(AnnotatedTypeFactory.java:3319)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getDeclAnnotations(AnnotatedTypeFactory.java:3299)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getDeclAnnotation(AnnotatedTypeFactory.java:3229)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getDeclAnnotation(AnnotatedTypeFactory.java:3164)
  org.checkerframework.framework.util.ContractsUtils.getPreconditions(ContractsUtils.java:245)
  org.checkerframework.framework.flow.CFAbstractTransfer.addInformationFromPreconditions(CFAbstractTransfer.java:518)
  org.checkerframework.framework.flow.CFAbstractTransfer.initialStore(CFAbstractTransfer.java:258)
  org.checkerframework.framework.flow.CFAbstractTransfer.initialStore(CFAbstractTransfer.java:93)
  org.checkerframework.dataflow.analysis.Analysis.init(Analysis.java:470)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:183)
  org.checkerframework.framework.flow.CFAbstractAnalysis.performAnalysis(CFAbstractAnalysis.java:91)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:1253)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:1171)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1520)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:258)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:320)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:167)
  jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
  jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:82)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:1012)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:522)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:182)
  jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:828)
  jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
  jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1418)
  jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1375)
  jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:973)
  jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:311)
  jdk.compiler/com.sun.tools.javac.main.Main.compile(Main.java:170)
  jdk.compiler/com.sun.tools.javac.Main.compile(Main.java:57)
  jdk.compiler/com.sun.tools.javac.Main.main(Main.java:43)
```
Used command: 
```
javac -cp lib/checker-2.11.0.jar -processor org.checkerframework.checker.nullness.NullnessChecker Handle.java
```
My java version:
```
java version "11.0.1" 2018-10-16 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.1+13-LTS)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.1+13-LTS, mixed mode)
```

### PR DESCRIPTION
**Pull request by smillst**
Fixes #2779.

### CODE REVIEW

**Comment by smillst** [comment]
The misc Azure job is failing, but the Travis one is passing.

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 1]
```diff
@@ -2254,26 +2254,22 @@ private boolean isSyntheticArgument(Tree tree) {
...
-                    AnnotatedTypeMirror ctxtype = TypeArgInferenceUtil.assignedTo(this, p);
-                    if (ctxtype != null) {
-                        fromNewClassContextHelper(type, ctxtype);
-                    } else {
-                        // give up trying and set to raw.
-                        type.setWasRaw();
-                    }
+                    // give up trying and set to raw.
+                    type.setWasRaw();
+                    type.getTypeArguments();
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: Is this call there for some side-effects? A comment would be useful.

File: framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 2]
```diff
@@ -2254,26 +2254,22 @@ private boolean isSyntheticArgument(Tree tree) {
...
-            if (((com.sun.tools.javac.code.Type) type.actualType)
-                    .tsym
-                    .getTypeParameters()
-                    .nonEmpty()) {
-                Pair<Tree, AnnotatedTypeMirror> ctx = this.visitorState.getAssignmentContext();
-                if (ctx != null) {
-                    AnnotatedTypeMirror ctxtype = ctx.second;
+            Pair<Tree, AnnotatedTypeMirror> ctx = this.visitorState.getAssignmentContext();
+            if (ctx != null && ctx.second.getKind() == TypeKind.DECLARED) {
+                AnnotatedTypeMirror ctxtype = ctx.second;
```
Comment (wmdietl) [inline framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: A comment here would be useful to understand why only a DECLARED context is used.

**Comment by abslamp** [comment]
https://github.com/typetools/checker-framework/pull/2807/files#diff-b997fe0effda8ef8635e7396f1861e64R2265-R2269
Do we really need this part? I tried to comment this out but allTests task still seems good.

**Comment by smillst** [comment]
@abslamp yes, it's faster to use the assignment context from the VisitorState if it exists.

**Comment by abslamp** [comment]
@smillst Using assignment context seems to have a problem with caching. If the caching is only disabled in dataflow, the nullness checker will crush when checking https://github.com/typetools/checker-framework/blob/master/framework/tests/all-systems/Issue2302.java. 
When visiting the return part in InitializationAnnotatedTypeFactory::visitNewClass (node=`new StrangeConstructorTypeArgs(new MyClass<>())`), the type of the assignment context is: `Pair(StrangeConstructorTypeArgs, @Initialized @NonNull StrangeConstructorTypeArgs<V extends @Initialized @Nullable Object>)`, which is not good enough for `MyClass<>()`. (But I suppose it is good enough for returning value?) By using the context, the type of `MyClass<>()` would be resolved into `MyClass<Object>` and cached, which would cause a crush later when checking the arguments, 
```
error: AnnotatedTypeMirrors aren't structurally equal.
    type1 = AnnotatedDeclaredType( @Initialized @NonNull Object )
    type2 = AnnotatedArrayType( @Initialized @NonNull byte @Initialized @NonNull [] )
    visitHistory = org.checkerframework.framework.type.StructuralEqualityVisitHistory@2a798d51
  Compilation unit: /home/pu/checker-framework/checker/tests/all-systems/Issue2302.java
  Last visited tree at line 17 column 16:
          return new StrangeConstructorTypeArgs(new MyClass<>());
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.javacutil.BugInCF.<init>(BugInCF.java:14)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.defaultAction(AbstractAtmComboVisitor.java:52)
  org.checkerframework.framework.type.StructuralEqualityComparer.defaultAction(StructuralEqualityComparer.java:68)
  org.checkerframework.framework.type.StructuralEqualityComparer.defaultAction(StructuralEqualityComparer.java:31)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visitDeclared_Array(AbstractAtmComboVisitor.java:123)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:303)
  org.checkerframework.framework.type.StructuralEqualityComparer.areEqual(StructuralEqualityComparer.java:110)
  org.checkerframework.framework.type.StructuralEqualityComparer.areEqualInHierarchy(StructuralEqualityComparer.java:122)
  org.checkerframework.framework.type.DefaultTypeHierarchy.areEqualInHierarchy(DefaultTypeHierarchy.java:308)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isContainedBy(DefaultTypeHierarchy.java:338)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitTypeArgs(DefaultTypeHierarchy.java:525)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:476)
  org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:41)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:307)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:182)
  org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:158)
  org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:2319)
  org.checkerframework.checker.nullness.NullnessVisitor.commonAssignmentCheck(NullnessVisitor.java:205)
  org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:2239)
  org.checkerframework.checker.nullness.NullnessVisitor.commonAssignmentCheck(NullnessVisitor.java:188)
  org.checkerframework.common.basetype.BaseTypeVisitor.checkArguments(BaseTypeVisitor.java:2815)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitNewClass(BaseTypeVisitor.java:1552)
  org.checkerframework.checker.nullness.NullnessVisitor.visitNewClass(NullnessVisitor.java:557)
  org.checkerframework.checker.nullness.NullnessVisitor.visitNewClass(NullnessVisitor.java:55)
  com.sun.tools.javac.tree.JCTree$JCNewClass.accept(JCTree.java:1705)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:301)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:167)
  com.sun.source.util.TreeScanner.visitReturn(TreeScanner.java:469)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitReturn(BaseTypeVisitor.java:1657)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitReturn(BaseTypeVisitor.java:167)
  com.sun.tools.javac.tree.JCTree$JCReturn.accept(JCTree.java:1548)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:301)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:167)
  com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
  com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:248)
  com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1026)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:301)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:167)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
  com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:206)
  org.checkerframework.framework.source.SourceVisitor.visitMethod(SourceVisitor.java:106)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitMethod(BaseTypeVisitor.java:687)
  org.checkerframework.checker.initialization.InitializationVisitor.visitMethod(InitializationVisitor.java:337)
  org.checkerframework.checker.initialization.InitializationVisitor.visitMethod(InitializationVisitor.java:52)
  com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:898)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
  org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:301)
  org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:167)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
  com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
  com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
  com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:187)
  org.checkerframework.framework.source.SourceVisitor.visitClass(SourceVisitor.java:94)
  org.checkerframework.common.basetype.BaseTypeVisitor.processClassTree(BaseTypeVisitor.java:388)
  org.checkerframework.checker.initialization.InitializationVisitor.processClassTree(InitializationVisitor.java:295)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:343)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:167)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:82)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:1020)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:522)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:182)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:828)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:120)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1404)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1363)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:959)
  com.sun.tools.javac.main.Main.compile(Main.java:302)
  com.sun.tools.javac.main.Main.compile(Main.java:162)
  com.sun.tools.javac.Main.compile(Main.java:57)
  com.sun.tools.javac.Main.main(Main.java:43)
1 error
```
I am wondering if the context should be updated when visiting the arguments.
Btw, it does not crush on interning and regex, which not using that method to visit new class.

**Pull request merged**